lib/python/qmk/cli/c2json.py

```diff
@@ -0,0 +1,324 @@
+from enum import auto, Flag
+import itertools
+import json
+import os
+from pathlib import Path
+from typing import List, NamedTuple
+
+from milc import cli
+import pycparser
+from pycparser.c_ast import BinaryOp, Constant, Decl, ID
+from pycparserext.ext_c_parser import GnuCParser
+
+
+def iter_upwards(path):
+    """
+    Generator to iterate through all parent directories yielding a Path
+    for each
+
+    Parameters:
+    path -- Path object to walk through
+    """
+    parts = list(path.resolve().parts)
+    while parts:
+        yield Path(*parts)
+        parts.pop()
+
+
+def find_topmost_keyboard_h(path):
+    """
+    Look through all directories up from the given path to find a keyboard.h
+    file. This is used for resolving QMK_KEYBOARD_H.
+
+    Parameters:
+    path -- path to the keymap.h we are resolving for
+
+    Returns:
+    A Path object for the keyboard.h found or None if we can't find one
+
+    Signature:
+    (path: Path) -> Path
+    """
+
+    def header_name(directory: Path) -> str:
+        return directory.name + '.h'
+
+    # path will look like
+    # keyboards/NAME/SUBNAME/keymaps/keymap.c
+    keyboard_h = None
+    for part in iter_upwards(path):
+        # if we hit a root level subdirectory we've overshot it
+        if part.name in ('keyboards', 'users', 'layouts'):
+            return keyboard_h
+
+        header = part / header_name(part)
+        if header.exists():
+            keyboard_h = header
+
+
+def get_qmk_root(from_path):
+    """
+    Finds the path of the QMK repository root relative to a keymaps directory
+
+    Parameters:
+    from_path -- Path to find the QMK root relative to
+
+    Signature:
+    (from_path: Path) -> Path
+    """
+    for part in iter_upwards(from_path.absolute()):
+        if (part / 'keyboards').exists():
+            return part
+
+
+def recursive_subdirs(from_path):
+    """
+    Generator that walks down the filesystem tree and gets all subdirectories
+    under the Path ``from_path``
+
+    Parameters:
+    from_path -- Path to walk from
+    """
+    yield from_path
+    for dirpath, dirs, files in os.walk(from_path):
+        yield from (Path(dirpath) / d for d in dirs)
+
+
+def parse_keymap_c(path):
+    """
+    Loads a C AST from the given keymap.c file.
+
+    Parameters:
+    path -- Path of the keymap.c to parse
+    """
+    qmk_root = get_qmk_root(path)
+    keyboard_h = find_topmost_keyboard_h(path)
+
+    # I'm guessing make probably doesn't just put the entire source tree as
+    # an include path??
```

Comment:
It doesn't, but discovering the current paths outside of make is... problematic. You can use `make VERBOSE=1 <keyboard>:<keymap>` to see the final list, but it may take some make trickery to get something that's futureproof.